### ISSUE TITLE
Applying "width of each Tab character" does not work

### ISSUE THREAD

**Issue by hoppfrosch** [Komodo/KomodoEdit#1543]
### Short Summary
Changing "width of each Tab character"-value (within Preferences->Editor->Indentation->Per Language Indentation Settings (Perl)) and pressing "Apply" button does not change the number of spaces inserted after pressing "TAG"	

### Steps to Reproduce
see Screenshot for my general indentation settings. Change "width of each Tab character"-value, press "Apply"  then "OK", and press "Tab" within Editor -> the old width-value is used instead of the new ... 

### Expected results
"Apply" should rely apply the changed value

### Actual results
"Apply" does not apply the changed value, but usese the old value

### Platform Information
Komodo IDE, version 10.0.1, build 89160, platform win32-x86.
Win7 SP1 Ultimate 64-bit 

![komodo_2016-05-19_09-56-46](https://cloud.githubusercontent.com/assets/419075/15386483/6ab1c02a-1da8-11e6-9da5-95a071c53a01.png)

**Comment by mitchell-as** [comment]
Thanks for the report.

**Comment by mitchell-as** [comment]
For now the workaround is to re-open the file, which will apply the new indentation settings.

**Comment by mitchell-as** [comment]
Moving to milestone 10.1 because this appears to have been a long-standing bug that will require significant effort to fix and may introduce too many regressions for a minor point release.

**Comment by mitchell-as** [comment]
Note to self: koDocuments only listen for global 'useTabs', 'indentWidth', and 'tabWidth' preference changes. They do not listen for language specific preference changes (hence the need to close and re-open the language-specific document). One potential fix is to connect language-specific preference observers based on a document's `_language` property (and disconnect when the language changes), and then on language-specific preference change, iterate through documents, looking for matching languages and changing them. Note however that the document's `_indentWidth` et. al. will not be changed. Anything depending on those old values will fail. This is another big bug.

**Comment by cgchoffman** [comment]
FWIW, I believe this is by design possibly because on the massive amount of work to implement this.

**Comment by mitchell-as** [comment]
If that's the case, then you've saved me a TON of work @cgchoffman :)

**Comment by cgchoffman** [comment]
I think if all you have to do is reopen the file then this a. shouldn't be priority and b. should rethink whether we actually do this.  What we could do is warn the user that these changes will not take effect on open files.  That warning can be found on other prefs I believe.  I like saving you work @mitchell-as !

**Comment by Naatan** [comment]
If the only reason not to do this is because it would take time then I don't think that's a good enough reason not to do it. Not to say we shouldn't give the warning as an interim solution.

**Comment by cgchoffman** [comment]
@Naatan I agree that only considering time would be inappropriate.  I'm saying time + added value still doesn't make sense when managing expectations (adding the warning) could be reasonable solution too.

@hoppfrosch, we're kind of talking around you.  Feel free to weigh in if you have thoughts.

**Comment by hoppfrosch** [comment]
As a user I would expect that Apply button applies the changed settings IMMEDIATELY with leaving the Preferences window open - whereas the OK-button applies all changed settings with closing the preferences box. At least that's what I'm used to from other applications - and that's what I would expect from komodo in an ideal case.

As it seems to be a too large effort in time to implement this, a viable solution might be, that pressing the Apply-button results in an dialog, which allows the user to choose whether the all the affected files should be reopened or not. Pressing yes, all affected windows (i.e. all "Perl" windows when changing the perl related indentation) are automatically reopened by Komodo. While doing so, the preferences dialog should stay open. Pressing OK should do offer the same reopen-behaviour - except the preferences window is closed after reopening the files.

**Comment by cgchoffman** [comment]
> As a user I would expect that Apply button applies the changed settings IMMEDIATELY with leaving the Preferences window open - whereas the OK-button applies all changed settings with closing the preferences box.

That is exactly how those buttons function.

> pressing the Apply-button results in an dialog, which allows the user to choose whether the all the affected files should be reopened or not.

This is a great idea.  It combines managing expectations with a viable workaround.  We should still include text by the preference input telling the user that it won't affect currently open buffers.

**Comment by Naatan** [comment]
Unfortunately that would likely be about as much work as properly fixing it, because reloading files isn't as simple as closing and opening them again (well you could make it that simple, but that would make it a very slow operation).

**Comment by cgchoffman** [comment]
The suggestion @hoppfrosch and I would be much less work then the proper fix (I'm personally not suggesting it as a proper fix, it's a stopgap).  It's telling Komodo to close then open files.  I acknowledge that this would be slower than propagating the preferences (which currently isn't an option).  Are you suggesting an alternative?  Or just saying "wait until we do a proper fix"?  If it's the latter then at the very least we need to add text to make it clear ot the user that the changes will not be reflected in the currently open files.

**Comment by Naatan** [comment]
Yeah I'm saying for the interim fix I prefer we keep it simple :) Cause reopening files could bring with it it's own set of problems.

**Comment by mitchell-as** [comment]
When I was working on #1246 , I added a notice about changing indentation settings on a per-language basis:

![1](https://cloud.githubusercontent.com/assets/12902397/16164655/64d50164-34ad-11e6-8505-e19c1d46a3c7.png)

Reading through this thread again, I'm not sure what else should be done. At present there is no way to "observe" (in preference-speak) a per-language indentation preference due to the fact a particular file can be any language. I am fine closing this ticket in favor of the fix to #1246 due to the fact that a user won't be fiddling with per-language indentation settings enough that re-opening files or restarting Komodo will become a nuisance.

There were so many little indentation problems in Komodo 9 and a few in 10.0.0 that I'm afraid of trying to implement basically a new feature since @cgchoffman mentioned the non-applying of per-language indentation settings immediately is "by design".

**Comment by Naatan** [comment]
I'll investigate this, to see if there is a way to reload files without affecting things like editor history.
